[NL Interface] Refactoring + Query parsing enhancement: remove stop words

jehangiramjad opened the pull request:
This is a fairly major enhancement. Today, we only remove the "place" detected (rather the part of the query which has the words for the place) from the query before doing SV detection. This means the query still has lots of stop words ("a", "and", "what about", "cities", "counties", "most", "best" etc). This has a major drawback for the embeddings matches: in the embeddings space, the semantic meaning can often get tripped up these extra words. 

As an example, in the current mechanism (without stop words removal), the following string:

"Which counties have agriculture workers in California" matches to the following top SVG/SV:

dc/g/Farm_FarmType-Cropland
Area_Farm_Cropland

Try on https://dev.datacommons.org/nl/

However, "Which counties have the most agriculture workers in California" matches to dc/hlxvn1t8b9bhh which is what we always want in this case. 

Asking for agriculture workers with the word "**counties**" makes things match closer to the "*_Cropland"  SVs which is not what one would expect. But "Which **cities** have agriculture workers in California" matches to dc/hlxvn1t8b9bhh. It appears that in the semantic space, including "counties" appears to bias the meaning towards "area" and "cropland". 

With the removal of these stop and classification detection words, we can do a lot better. All of the queries above now bring up dc/hlxvn1t8b9bhh  as the top match.

I have also tested all Scenarios in the demo script and they both work 100% the same as the case without the stopwords removal.

shifucun commented on server/services/nl.py:
```diff
@@ -48,20 +48,134 @@
 MODEL_NAME = 'all-MiniLM-L6-v2'
 
 STOP_WORDS = {
```
Can we move these constants and corresponding functions to one file in server/lib/nl_stop_words.py?

I would highly recommend to create a new folder: `server/lib/nl` and put current nl_*.py there too, so `lib/nl_data_spec.py` becomes `lib/nl/data_spec.py`

jehangiramjad replied:
I moved some constants to lib/nl/nl_constants.py and some processing to lib/nl/nl_utils.py

shifucun commented on server/services/nl.py:
```diff
@@ -120,11 +234,38 @@
 }
 
 
+def _add_to_set_from_list(set_strings, list_string):
```
would be nice to start adding tests for these util functions, then we would expand as we see issues from demo.

jehangiramjad replied:
Done for some of the functions introduced in this PR

juliawu commented on server/services/nl.py:
```diff
@@ -120,11 +234,38 @@
+STOP_WORDS_QUERY_CLASSIFICATION = set()
+for (_, v) in QUERY_CLASSIFICATION_HEURISTICS.items():
+  if isinstance(v, list):
+    # If 'v' is a list, add all the words.
```
Some of the entries of v might be multiple word phrases. Either we should rework the entries of v to be one word only, or we should add logic to remove multi word phrases first

jehangiramjad replied:
Good point! After the refactoring, I am now inserting words after splitting.

juliawu reviewed (commented):
Very cool!

jehangiramjad commented:
Did some refactoring to now have lib/nl/nl_utils.py and lib/nl/nl_constants.py Also updated the various dependencies due to the changing paths for several files. Additionally, now added some unit tests for functions in nl_utils.py

In future commits/PRs, we should do a similar refactoring for:

1. Place detection (currently all in services/nl.py)
2. SV detection (currently all in services/nl.py)
3. Query classification (currently all in services/nl.py)

juliawu reviewed (approved):
Thanks for the changes, I really like the refactoring!

shifucun commented on server/lib/nl/__init__.py:
```diff
@@ -0,0 +1,13 @@
+# Copyright 2020 Google LLC
```
2023

jehangiramjad replied:
Done, thanks!

shifucun reviewed (approved):
Thanks for the refactor, looks so much better!
